### PR TITLE
resolve aliases in subqueries in function arguments

### PR DESCRIPTION
**Pull request by JCOR11599**
The rule `reorderProjection` also replaces subqueries with getfields in projections when they are used by subqueries, but it did not check for function expressions.

This meant that aliases in subqueries as arguments to functions threw a `"x" could not be found` error.
This PR just has the section of `reorderProjection` that is supposed to find deferredColumns also look at the arguments of functions recursively (because we can nest functions).

Additionally, there was another schema type bug:
```sql
tmp> select 0 as foo, if((select foo), 123, 456);
+-----+----------------------------+
| foo | if((select foo), 123, 456) |
+-----+----------------------------+
| 0   | 127                        |
+-----+----------------------------+
1 row in set (0.00 sec)
```
MySQL returns an Integer type for if statement, and if either argument is a String, it always returns a String.

fix for: https://github.com/dolthub/dolt/issues/6174

### CODE REVIEW

File: enginetest/queries/queries.go [hunk 1]
```diff
@@ -5142,6 +5142,42 @@ Select * from (
...
+	{
+		Query: `SELECT if(0, 123, 456)`,
+		Expected: []sql.Row{
+			{456},
+		},
+	},
+	{
+		Query: `SELECT if(0, "abc", 456)`,
+		Expected: []sql.Row{
+			{456},
```
Comment (JCOR11599) [inline enginetest/queries/queries.go:5160]: This should be a string, but the TestEvaluation doesn't go through the schema[0].Convert code path.

Comment (max-hoffman) [reply enginetest/queries/queries.go:5160]: should we move to brokenQueries?

Comment (max-hoffman) [reply enginetest/queries/queries.go:5160]: or is it just bad test setup

File: sql/analyzer/reorder_projections.go [hunk 2]
```diff
@@ -123,6 +123,30 @@ func reorderProjection(ctx *sql.Context, a *Analyzer, n sql.Node, scope *plan.Sc
 	})
 }
 
+// findDeferredColumns is a function that searches through a projection's expressions and checks if there are any
+// subqueries that reference the deferred column. Additionally, we have to check for subqueries in functions arguments
+func findDeferredColumns(exprs ...sql.Expression) []column {
```
Comment (max-hoffman) [inline sql/analyzer/reorder_projections.go:128]: is there any reason not to just `transform.Expr` for each of these?

File: server/handler_test.go [hunk 3]
```diff
@@ -188,6 +188,28 @@ func TestHandlerOutput(t *testing.T) {
 		require.Equal(t, sqltypes.Float64, result.Rows[0][0].Type())
 		require.Equal(t, []byte("1"), result.Rows[0][0].ToBytes())
 	})
+
+	t.Run("if() type is correct", func(t *testing.T) {
+		handler.ComInitDB(dummyConn, "test")
+		var result *sqltypes.Result
+		err := handler.ComQuery(dummyConn, "select if(1, 123, 'def')", func(res *sqltypes.Result, more bool) error {
```
Comment (max-hoffman) [inline server/handler_test.go:195]: should be do `(0, "def", 123)` -> `"123`?

**Review by max-hoffman — APPROVED** [review]
LGTM, could probably add more tests, but we need a dedicated expression typing module before realistically fixing the long tail

File: enginetest/queries/queries.go (hunk 1 above)
Comment (JCOR11599) [reply enginetest/queries/queries.go:5160]: I think it's just bad test setup, when the results are written to the shell we use the writerIters, but not for evaluation.
I can add it in a separate PR, and see how many failures we get

File: server/handler_test.go (hunk 3 above)
Comment (JCOR11599) [reply server/handler_test.go:195]: The old code had `return n.ifTrue.Type()`, which was already Text, so it didn't actually test the new code I added.

**Pull request merged**